drivers/mailbox/bflb-ipc.c

```diff
@@ -157,17 +192,42 @@ static const struct irq_domain_ops bflb_ipc_irq_ops = {
 	.xlate = bflb_ipc_domain_xlate,
 };
 
+/* JH: Figure out if M0 has processed the last signal sent
+ * by checking if the High/Low registers are cleared
+ */
+static int bflb_ipc_mbox_can_send(struct mbox_chan *chan)
+{
+	struct bflb_ipc *ipc = to_bflb_ipc(chan->mbox);
+
+	u32 mbox_high = readl(ipc->base[2] + IPC_REG_ILSHR);
+	u32 mbox_low = readl(ipc->base[2] + IPC_REG_ILSLR);
+
+	dev_dbg(ipc->dev, "%s: low: 0x%08x high: 0x%08x\r\n", __func__, mbox_low, mbox_high);
+
+	return !(mbox_low | mbox_high);
+}
+
+
+
 static int bflb_ipc_mbox_send_data(struct mbox_chan *chan, void *data)
 {
 	struct bflb_ipc *ipc = to_bflb_ipc(chan->mbox);
 	struct bflb_ipc_chan_info *mchan = chan->con_priv;
-	u32 hwirq;
 
-	hwirq = bflb_ipc_get_hwirq(mchan->client_id, mchan->signal_id);
+	if (!bflb_ipc_mbox_can_send(chan))
+		return -EBUSY;
 
-	dev_dbg(ipc->dev, "%s: hwirq: %u\n", __func__, hwirq);
 
-//	writel(hwirq, ipc->base + IPC_REG_SEND_ID);
+	dev_dbg(ipc->dev, "%s: %d %d\n", __func__, mchan->client_id, mchan->signal_id);
+
+
+	// /* write our signal number to high register */
+	writel(mchan->signal_id, ipc->base[2] + IPC_REG_ILSHR);
```

Comment:
I was thinking to allocate some bits for client number, and some bits for signal number, then only the global list of clients has to be kept in sync between drivers, and each driver can allocate signals however they want.  This would likely require expanding the interrupt-cells in the device-tree so that both can be specified, and changing BFLB_IPC_DEVICE to now be "client id"

Reply:
#interupt-cells or #mbox-cells?
and how would we handle LP IPC (when it comes?). I understand changing the base register but where would we specify that in the mailbox?